prototype/jsonnet_snippet.go

```
@@ -0,0 +1,222 @@
+// Copyright 2017 The kubecfg authors
+//
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+package prototype
+
+import (
+	"sort"
+	"strings"
+
+	ast "github.com/google/go-jsonnet/ast"
+	parser "github.com/google/go-jsonnet/parser"
+
+	"github.com/ksonnet/kubecfg/prototype/snippet"
+)
+
+const (
+	paramPrefix            = "param://"
+	paramReplacementPrefix = "${"
+	paramReplacementSuffix = "}"
+)
+
+// Parse rewrites the imports in a Jsonnet file before returning the parsed
+// TextMate snippet.
+func Parse(fn string, jsonnet string) (snippet.Template, error) {
+	s, err := parse(fn, jsonnet)
+	if err != nil {
+		return nil, err
+	}
+
+	return snippet.Parse(s), nil
+}
+
+func parse(fn string, jsonnet string) (string, error) {
+	tokens, err := parser.Lex(fn, jsonnet)
+	if err != nil {
+		return "", err
+	}
+
+	root, err := parser.Parse(tokens)
+	if err != nil {
+		return "", err
+	}
+
+	var imports []ast.Import
+
+	// Gather all parameter imports
+	visit(root, &imports)
+
+	// Replace all parameter imports
+	return replace(jsonnet, imports), nil
+}
+
+// ---------------------------------------------------------------------------
+
+func visit(node ast.Node, imports *[]ast.Import) {
+	switch n := node.(type) {
+	case *ast.Import:
```

Hmm, I think we're missing:

* `ImportStr`
* `Dollar`
* `LiteralBoolean`
* `LiteralNull`
* `LiteralNumber`
* `LiteralString`
* `Self`
* `Var`

I think this is ok, (since none of them have members we need to visit, but I think we should also explicitly decide whether we want to error out if we find a node with an unrecognized type.

My suspicion is yes, since if we add syntax to Jsonnet (as we did recently for `SuperIn`), then it will be possible for a user to write `import "params://"` and have it not picked up, because we don't know how to visit the children.

So, I recommend that we make the above nodes explicitly handled by the `switch`, but do nothing when we visit them. Then, we might have a `default` that errors out if it doesn't recognize the node type.

Thoughts?

---

This makes sense to me, as we do probably want to error out for invalid Jsonnet input.